src/Compilers/CSharp/Test/Emit3/Semantics/ExtensionTests.cs

```diff
@@ -31210,4 +31326,1317 @@ static Expression<Func<Func<string, string>>> Test()
+            //         public async void M2() { await System.Threading.Tasks.Task.Yield(); }
+            Diagnostic(ErrorCode.ERR_BadSpecialByRefParameter, "M2").WithArguments("System.Span<int>").WithLocation(9, 27));
+    }
+
+    [Fact]
+    public void ReceiverParameterValidation_Iterator_01()
+    {
+        string source = """
+static class E
+{
+    extension(ref int i)
+    {
+        public static System.Collections.Generic.IEnumerator<int> M() { yield return 0; }
+    }
+    extension(ref int i)
+    {
+        public System.Collections.Generic.IEnumerator<int> M2() { yield return 0; }
+    }
+}
+""";
+        var comp = CreateCompilation(source);
+        comp.VerifyEmitDiagnostics(
+            // (9,60): error CS1623: Iterators cannot have ref, in or out parameters
+            //         public System.Collections.Generic.IEnumerator<int> M2() { yield return 0; }
+            Diagnostic(ErrorCode.ERR_BadIteratorArgType, "M2").WithLocation(9, 60));
+    }
+
+    [Fact]
+    public void ReceiverParameterValidation_Iterator_02()
+    {
+        string source = """
+unsafe static class E
+{
+    extension(int* i)
+    {
+        public static System.Collections.Generic.IEnumerator<int> M() { yield return 0; }
+    }
+    extension(int* i)
+    {
+        public System.Collections.Generic.IEnumerator<int> M2() { yield return 0; }
+    }
+}
+""";
+        var comp = CreateCompilation(source, options: TestOptions.UnsafeDebugDll);
+        comp.VerifyEmitDiagnostics(
+            // (3,15): error CS1103: The receiver parameter of an extension cannot be of type 'int*'
+            //     extension(int* i)
+            Diagnostic(ErrorCode.ERR_BadTypeforThis, "int*").WithArguments("int*").WithLocation(3, 15),
+            // (7,15): error CS1103: The receiver parameter of an extension cannot be of type 'int*'
+            //     extension(int* i)
+            Diagnostic(ErrorCode.ERR_BadTypeforThis, "int*").WithArguments("int*").WithLocation(7, 15));
+    }
+
+    [Fact]
+    public void ReceiverParameterValidation_CancellationTokenParameter_Static()
+    {
+        string source = """
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+static class C
+{
+    extension([EnumeratorCancellation] CancellationToken token1)
+    {
+        static async System.Collections.Generic.IAsyncEnumerable<int> Iter(int value, [EnumeratorCancellation] CancellationToken token2)
+        {
+            _ = token2.IsCancellationRequested;
+            yield return value++;
+            await Task.Yield();
+        }
+    }
+}
+""";
+        var comp = CreateCompilation(source, targetFramework: TargetFramework.Net90);
+        comp.VerifyDiagnostics(
+            // (7,16): warning CS8424: The EnumeratorCancellationAttribute applied to parameter 'token1' will have no effect. The attribute is only effective on a parameter of type CancellationToken in an async-iterator method returning IAsyncEnumerable
```

Comment:
>warning CS8424: The EnumeratorCancellationAttribute applied to parameter 'token1' will have no effect. [](http://example.com/codeflow?start=23&length=102)

Given the speclet wording "The _receiver parameter_ may not have an `[EnumeratorCancellation]` attribute." I would expect an error instead.

Reply:
Thanks. I'll fix the spec